server/src/main/scala/org/http4s/server/middleware/Caching.scala

```diff
@@ -0,0 +1,111 @@
+package org.http4s.server.middleware
+
+import cats._
+import cats.implicits._
+import cats.effect._
+import cats.data._
+import org.http4s._
+import org.http4s.util.CaseInsensitiveString
+import org.http4s.headers.{Date => HDate, _}
+import scala.concurrent.duration._
+
+object Caching {
+  def `no-store`[G[_]: Monad: Clock, F[_], A](
+      http: Kleisli[G, A, Response[F]]): Kleisli[G, A, Response[F]] =
+    Kleisli { a: A =>
+      for {
+        resp <- http(a)
+        now <- HttpDate.current[G]
+      } yield {
+        val headers = List(
+          `Cache-Control`(
+            NonEmptyList.of[CacheDirective](
+              CacheDirective.`no-store`,
+              CacheDirective.`private`(),
+              CacheDirective.`no-cache`(),
+              CacheDirective.`max-age`(0.seconds)
+            )),
+          Header("Pragma", "no-cache"),
+          HDate(now),
+          Expires(HttpDate.Epoch) // Expire at the epoch for no time confusion
+        )
+        resp.putHeaders(headers: _*)
+      }
+    }
+
+  object Helpers {
+    def defaultStatussToSetOn(s: Status): Boolean =
+      s match {
+        case Status.NotModified => true
+        case otherwise => otherwise.isSuccess
+      }
+
+    def defaultMethodsToSetOn(m: Method): Boolean = methodsToSetOn.contains(m)
+
+    private lazy val methodsToSetOn: Set[Method] = Set(
+      Method.GET,
+      Method.HEAD
+    )
+  }
+
+  def publicCache[G[_]: MonadError[*[_], Throwable]: Clock, F[_]](
+      lifetime: Duration,
+      http: Http[G, F]): Http[G, F] =
+    cache(
+      lifetime,
+      Either.left(CacheDirective.public),
+      Helpers.defaultMethodsToSetOn,
+      Helpers.defaultStatussToSetOn,
+      http)
+
+  def privateCache[G[_]: MonadError[*[_], Throwable]: Clock, F[_]](
+      lifetime: Duration,
+      http: Http[G, F],
+      fieldNames: List[CaseInsensitiveString] = Nil): Http[G, F] =
+    cache(
+      lifetime,
+      Either.right(CacheDirective.`private`(fieldNames)),
+      Helpers.defaultMethodsToSetOn,
+      Helpers.defaultStatussToSetOn,
+      http)
+
+  def cache[G[_]: MonadError[*[_], Throwable]: Clock, F[_]](
+      lifetime: Duration,
+      isPublic: Either[CacheDirective.public.type, CacheDirective.`private`],
+      methodToSetOn: Method => Boolean,
+      statusToSetOn: Status => Boolean,
+      http: Http[G, F]
+  ): Http[G, F] = {
+    val actualLifetime = lifetime match {
+      case finite: FiniteDuration => finite
+      case _ => 315360000.seconds // 10 years
+      // Http1 caches do not respect max-age headers, so to work globally it is recommended
+      // to explicitly set an Expire which requires some time interval to work
```

Comment:
Nice comment.  I was getting ready to throw a fit about `Duration` about 20 lines above.